tests/virtualization/universal/waitfor_guests.pm

```diff
@@ -15,118 +15,38 @@ use warnings;
-
-        # Apply xml settings to VM. Note: They will be applied after reboot.
-        assert_script_run("virt-xml-validate $xml");
-        assert_script_run("virsh define $xml");
-
-    } elsif (is_xen_host()) {
-        # We're skipping to add an additional bridge on xen
-    } else {
-        my $msg = "Unknown machine type";
-        record_soft_failure($msg);
-    }
-}
-
-# Upload xml definitions of all guests
-sub upload_machine_definitions {
-    foreach my $guest (keys %virt_autotest::common::guests) {
-        my $xml = "$guest.xml";
-        assert_script_run("virsh dumpxml $guest | tee $xml");
-        upload_logs("$xml");
-    }
-}
-
 sub run {
     my $self = shift;
     select_console('root-console');
+    my @guests = keys %virt_autotest::common::guests;
     # Fill the current pairs of hostname & address into /etc/hosts file
     assert_script_run 'virsh list --all';
-    add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (keys %virt_autotest::common::guests);
+    add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (@guests);
     assert_script_run "cat /etc/hosts";
 
     # Wait for guests to announce that installation is complete
     script_retry("test -d /tmp/guests_ip", retry => 15, delay => 120);
-    foreach my $guest (keys %virt_autotest::common::guests) {
+    foreach my $guest (@guests) {
         script_retry("test -f /tmp/guests_ip/$guest", retry => 20, delay => 120);
         record_info("$guest installed", "Guest installation completed");
     }
     record_info("All guests installed", "Guest installation completed");
-
-    # Adding the PCI bridges requires the guests to be shutdown
-    record_info("shutdown guests", "Shutting down all guests");
-    shutdown_guests();
-
-    ## Add a PCIe root port and a PCIe to PCI bridge for hotplugging
-    add_pci_bridge("$_") foreach (keys %virt_autotest::common::guests);
-
-    ## Ensure the reboot policy is set to 'restart'. This needs to happen on shutdown guest
-    # Do a shutdown and start here because some guests might not reboot because of the on_reboot=destroy policy
-    ensure_reboot_policy("$_") foreach (keys %virt_autotest::common::guests);
-    upload_machine_definitions();
-
-    record_info("Starting guests", "Starting all guests");
-    start_guests();
-
-    # Check that guests are online so we can continue and setup them
-    ensure_online $_, skip_ssh => 1, ping_delay => 45 foreach (keys %virt_autotest::common::guests);
-
-    # All guests should be now installed and running
+    my $host_version = get_var("VERSION") =~ s/-SP/./r;
+    if ($host_version > 15 && get_var("KVM")) {
+        # Adding the PCI bridges requires the guests to be shutdown
+        record_info("shutdown guests", "Shutting down all guests");
+        shutdown_guests();
+
+        ## Add a PCIe root port and a PCIe to PCI bridge for Q35 machine
+        if ($host_version < 15.4) {
```

Comment:
is_sle('<15-SP4') ?

Reply:
Yes, 15-SP4 has 14 pcie ports by default while earlier versions have 7  ports. 15-SP4 have enough spare ports to run our tests so  there is no need add.